code/data_structures/src/heaps/binomial_heap/binomial_heap.hpp

```
@@ -0,0 +1,609 @@
+	 * 		node is the root node of a heap
+	 */
+	binomialNode<T> * head;
+};
+
+
+template <typename T>
+binomialHeap<T>::binomialHeap()
+{
+	head = 0;
+}
+
+template <typename T>
+binomialHeap<T>::~binomialHeap()
+{
+	if (head)
+		delete head;
+}
+
+template <typename T>
+bool binomialHeap<T>::isEmpty()
+{
+	return (head==0);
+}
+
+
+template <typename T>
+void binomialHeap<T>::insert(T v)
+{
+	binomialNode<T>*  p = head?  head->parent:  0;
+	binomialNode<T> * singleton = new binomialNode<T>(v, p);
+	head = unionHeap(singleton, head);
+
+}
+
+template <typename T>
+T binomialHeap<T>::getMin()
+{
+	T min;
+	if(head)
+	{
+		binomialNode<T> * temp = head;
+		min = head->value;
+		while(temp)
+		{
+			if(temp->value < min)
+			{
+				min = temp->value;
+			}
+			temp = temp->sibling;
+		}
+
+	}
+	return min;
+}
+template <typename T>
+T binomialHeap<T>::extractMin()
+{
+	binomialNode<T>*prev=0, *min=0, *temp=0, *next=0;
+	T minValue;
+	if(head)
+	{
+		binomialNode<T>* minPrev = 0; //previous node of min node
+		min = head;
+		temp = head->sibling;
+		prev= head;
+		/*
+		 * Find min and minPrev of heaps
+		 */
+		while(temp)
+		{
+			if(temp->value < min->value)
+			{
+				min = temp;
+				minPrev= prev;
+			}
+			prev= prev->sibling;
+			temp = temp->sibling;
+		}
+		/*
+		 *  If prev, assign prev sibling to min sibling.
+		 *  Else if no prev min->sibling is new head
+		 */
+		if(minPrev)
+		{
+			minPrev->sibling=min->sibling;
+		}
+		else
+		{
+			head = min->sibling;
+		}
+		/*
+		 * Update all children nodes parent pointers to null
+		*/
+		next = min->child;
+		temp = next;
+		while(temp)
+		{
+			temp->parent = min->parent;
+			temp = temp->sibling;
+		}
+		/*
+		 * Detach and delete min
+		 */
+		min->sibling = 0;
+		min->child = 0;
+		min->parent = 0;
+		minValue=  min->value;
+		delete min;
+		/*
+		 * Union the two detached heaps
+		 */
+		head = unionHeap(head, next);
+		temp = min = minPrev = next= 0;
+	}
+	else
```

Better to throw an exception